### PR TITLE
Leaves - Kristina

### PR DESCRIPTION
**Pull request by krismosk**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a custom method that would select the top 10 works with the highest votes for each category.
Describe how you approached testing that model method. What edge cases did you come up with? | I chose to test for if the method returned no works of that type, that it would return nil. I also tested if it returned less than 10 works and more than 10 works.
What are `session` and `flash`? What is the difference between them? | Flash is a special hash that holds a message from the controller that it sends to the view. A flash message is used a single time, and is displayed once before it goes away. The message is available for the request cycle and the next complete request cycle. A session is a special hash object hat holds data during a user's session. It is similar to flash, but the data inside session will not disappear after the next request-response cycle and will be stored indefinitely. 
What was one thing that you gained more clarity on through this assignment? | Databases! This project really helped me solidify the concepts that I was introduced with the Rideshare Rails project last week. I feel more comfortable with writing seed scripts, seeding databases, foreign keys, and model relationships.
What is the Heroku URL of your deployed application? | https://kristina-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | I would suggest giving the students more flexibility on the design of the website so that we can be more creative instead of copying the same styling as the mock website.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  def logout
+    flash[:success] = "Successfully logged out."
+    session[:user_id] = nil
+    redirect_to root_path
+  end
+
+
+  private
+
+  def find_user
```
Comment (beccaelenzil) [inline app/controllers/users_controller.rb:64]: Consider putting this method in your application controller so other controllers will also inherit this method.

File: test/models/vote_test.rb [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    it "must have a work" do
+      @vote.work_id = nil
+      expect(@vote.valid?).must_equal false
+      expect(@vote.errors.messages).must_include :work
+      expect(@vote.errors.messages[:work]).must_equal ["must exist", "can't be blank"]
+    end
+  end
+
+  describe "relationships" do
```
Comment (beccaelenzil) [inline test/models/vote_test.rb:90]: These tests look great. Consider whether it's possible to use your fixtures for these tests.

**Comment by beccaelenzil** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully |yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | yes
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | becca
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI |
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works | yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | see comment
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category |yes 
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | yes
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) |yes 
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback


Great job overall! Your implementation matches the demo site very closely, and the learning goals for this assignment were definitely met. Your code is well written and well tested. Keep up the hard work!

File: app/views/works/home.html.erb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    <h2 class="spotlight__header">
+      <% spotlight_work = Work.select_spotlight%>
+      <span class="spotlight__header--prefix">Media Spotlight</span>
+      <span class="spotlight__link-to"><%= link_to spotlight_work.title, work_path(spotlight_work.id) %></span>
+      <%= "by: #{spotlight_work.creator}"%>
+    </h2>
+    <p><%= spotlight_work.votes.length %> votes</p>
+  </section>
+
+  <section class="top-ten__container">
```
Comment (beccaelenzil) [inline app/views/works/home.html.erb:12]: Consider using partial view to DRY up repeated code
